src/file_io.jl

```diff
@@ -107,3 +125,26 @@ end
 # FileIO Interface
 fileio_save(f, args...; kwargs...) = save(f.filename, args...; kwargs...)
 fileio_load(f, args...) = load(f.filename, args...)
+
+# Objects were initialially saved inline with the BSON document prior to v4
+# To support larger objects/files we started writing objects to the end of file and only save
+# names and sizes to the objects dict in the bson doc.
+# Since our ideal structure is one dict with object bytes we don't upgrade, but rather
+# extract both formats.
+function _extract_objects(io::IO, d)
```

Comment:
Should we make some note about the required state of the `io` object?
I.e. that pre-v4 it should be at end of file,
and for v4, it should have had the BSON metadata header already read off of it, and thus be at the start of the data section?

Reply:
Hmmm, I wonder if it would be safer to include a `object-pos` array which specifies a position to seek to in the file io stream. In julia, if you seek to an invalid position and try to read then it'll just throw an `EOFError` anyways. This would also be helpful for libraries that don't support an IO based interface. For example, `bson.loads` correctly reads the header, but you're still stuck reading the first 4 BSON bytes to figure out the file offset to start reading the raw data from.

Reply:
Opted to just include a positions array as well, so the io state isn't an externality to this function.